### PR TITLE
Implement keyboard interactions with DeviceAgent

### PR DESCRIPTION
**Pull request by jmoody**
### Motivation

Completes:

* Implement text entry with DeviceAgent [JIRA](https://xamarin.atlassian.net/browse/TCFW-240)
* Remove iPad keyboard modes support from Calabash iOS [JIRA](https://xamarin.atlassian.net/browse/TCFW-340)
* Keyboard action key for "Join" button has different identifiers for devices and simulators [JIRA](https://xamarin.atlassian.net/browse/TCFW-361)

Progress on:

* Calabash iOS DeviceAgent action layer [JIRA](https://xamarin.atlassian.net/browse/TCFW-31)
* iOS returnKeyType "Continue" does not have stable identifier [JIRA](https://xamarin.atlassian.net/browse/TCFW-344)

### Breaking Changes

The next release of Calabash will be 0.20.0 and we are introducing breaking changes. 

1. The `POST_ENTER_KEYBOARD` is no longer recognized by Calabash.
2. When `keyboard_enter_char` is passed "Return" or "Delete", it will attempt to type the
   corresponding escape character (\n and \b respectively).  This is a breaking change.
   Previously, these arguments called out to methods that attempted to tap keyboard keys.
   You should use the following methods instead:
   * `tap_keyboard_action_key`
   * `tap_keyboard_delete_key` (NEW)
3. The following methods were used on iPads to force the keyboard into a given mode.  These
   methods stopped working in iOS 9 and were never fixed.  They have been removed.
   * `ensure_split_keyboard`
   * `ensure_docked_keyboard`
   * `ensure_undocked_keyboard`

### New API

* `wait_for_no_keyboard` (2.x compatible!)
* `tap_keyboard_delete_key`
* `enter_text_in` (alias of `enter_text` - 2.x compatible!)
* `text_from_first_responder` (`_text_from_first_responder` remains)

### Notes

`enter_text` and `fast_enter_text` were added to Calabash in 0.11.2.  These methods tried to make text entry less painful by using UIAutomation `setValue()`.  The text entry with DeviceAgent is really fast and (so far) reliable, so there are no special implementations for these methods - they will always use the generic text entry capabilities of DeviceAgent.

`enter_text` corresponds to `enter_text_in` in Calabash 2.0.  As such, the `enter_text` method has been renamed to `enter_text_in`.  There is a method alias: `enter_text`. You should start using `enter_text_in` to be (more) 2.0 compatible.

### Test

```
# If you use bundler for local gem development (you know who you are).
$ cd path/to/run_loop
$ git co develop
$ git pull

# Get this branch
$ git co -t origin/feature/implement-text-entry-with-DeviceAgent
$ cd calabassh-cucumber/test/cucumber

# Run against instruments
$ be cucumber -t @keyboard

# Run against device_agent
$ CBX_LAUNCHER=ios_device_manager be cucumber -t @keyboard
```

Tested against iOS 9 and 10 physical devices with Xcode 7 and 8 using iOSDeviceManager _and_ xcodebuild as the launcher.

I ran this against CalSmokeApp and the @keyboard tag - some failures related to "Shift" key and autocorrect - otherwise all good!

### CODE REVIEW

**Comment by jmoody** [comment]
@jescriba Sorry, this is a big one.  Let's sync tomorrow if it is too much to digest.

The good news is that we have ~90% coverage on the new and existing keyboard APIs.

File: calabash-cucumber/lib/calabash-cucumber/core.rb [hunk 1]
```diff
@@ -462,6 +465,192 @@ def pinch(in_out, options={})
...
+      end
+
+      # Uses the keyboard to enter text.
+      #
+      # @param [String] text the text to type.
+      # @raise [RuntimeError] If the keyboard is not visible.
+      def keyboard_enter_text(text)
+        expect_keyboard_visible!
+        existing_text = text_from_first_responder
+        if existing_text && existing_text == ""
```
Comment (jescriba) [inline calabash-cucumber/lib/calabash-cucumber/core.rb]: Do we need this if/else? Seems like `text_from_first_responder` cannot return nil either way and the gsub should leave the string unaltered if existing_text = ""

Comment (jmoody) [inline calabash-cucumber/lib/calabash-cucumber/core.rb]: Ah!  Right.  This is an artifact from a time long ago when `text_from_first_responder` returned `nil`.

Can you make the change?  Or do you you want me to?

Comment (jmoody) [inline calabash-cucumber/lib/calabash-cucumber/core.rb]: But don't forget to run the rspec tests!  I recommend running `guard` in another Terminal.

**Pull request merged**